File: ios/Halogen/Components/ListControls.swift
import SwiftUI

/// Sort field vocabulary for episode lists (server `order[order_by]` names).
/// `.position` is the queue's pivot order (local, queue-only).
enum EpisodeOrderField: String, Codable, CaseIterable, Identifiable {
    case published = "published_at"
    case added = "created_at"
    case title
    case duration = "duration_secs"
    case position
    /// History only: play-recency order, local-only. A dedicated case so an
    /// EXPLICIT "Published desc" choice is distinguishable from the recency default.
    case recency

    var id: String { rawValue }

    var label: String {
        switch self {
        case .published: return "Published"
        case .added: return "Added"
        case .title: return "Title"
        case .duration: return "Duration"
        case .position: return "Position"
        case .recency: return "Recent"
        }
    }
}

/// One episode list's controls state: search + filter chips + order.
/// Persisted per page (the web's per-view-key sticky list config). Chips are
/// multi-select: OR within a facet, AND across facets (web query.rs).
struct ListQuery: Codable, Equatable {
    var search: String = ""
    var filters: Set<EpisodeFilter> = []
    var orderField: EpisodeOrderField = .published
    var direction: OrderDirection = .desc

    private enum CodingKeys: String, CodingKey {
        case search, filters, orderField, direction
    }

    init(
        search: String = "", filters: Set<EpisodeFilter> = [],
        orderField: EpisodeOrderField = .published, direction: OrderDirection = .desc
    ) {
        self.search = search
        self.filters = filters
        self.orderField = orderField
        self.direction = direction
    }

    /// Lenient decode: pre-multiselect snapshots (single `filter` key) just
    /// reset to no chips — never fail the whole restore over a vocabulary
    /// change.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        search = (try? c.decode(String.self, forKey: .search)) ?? ""
        filters = (try? c.decode(Set<EpisodeFilter>.self, forKey: .filters)) ?? []
        orderField = (try? c.decode(EpisodeOrderField.self, forKey: .orderField)) ?? .published
        direction = (try? c.decode(OrderDirection.self, forKey: .direction)) ?? .desc
    }

    /// Selected chips of each facet.
    var downloadChips: Set<EpisodeFilter> { filters.intersection(EpisodeFilter.downloadFacet) }
    var playedChips: Set<EpisodeFilter> { filters.intersection(EpisodeFilter.playedFacet) }

    /// Whether the server page must be re-filtered locally: a facet with
    /// MULTIPLE chips can't be expressed as a single wire token, so the fetch
    /// goes un-narrowed for that facet and `matchesChips` trims each page.
    var needsLocalChipFilter: Bool {
        downloadChips.count > 1 || playedChips.count > 1
    }

    /// Server-side query fragment (search + single-chip facets + order).
    /// Queue/History apply the same query locally instead.
    var queryItems: [URLQueryItem] {
        var items: [URLQueryItem] = []
        let trimmed = search.trimmingCharacters(in: .whitespaces)
        if !trimmed.isEmpty {
            items.append(URLQueryItem(name: "filter[search]", value: trimmed))
        }
        if downloadChips.count == 1, let token = downloadChips.first?.wireToken {
            items.append(URLQueryItem(name: "filter[download_status]", value: token))
        }
        if playedChips.count == 1, let token = playedChips.first?.wireToken {
            items.append(URLQueryItem(name: "filter[playback_status]", value: token))
        }
        if orderField != .position, orderField != .recency {
            items.append(URLQueryItem(name: "order[order_by]", value: orderField.rawValue))
            items.append(
                URLQueryItem(name: "order[direction]", value: direction == .asc ? "Asc" : "Desc"))
        }
        return items
    }

    /// Chip predicate over one row: OR within a facet, AND across facets (web
    /// apply_filter_sort). `isOnDevice` resolves the OnDevice chip (nil = no-op
    /// when callers switch to the device set wholesale); `status` should be the
    /// overlay read so a local mark-played flips results immediately.
    func matchesChips(
        _ episode: EpisodeData,
        isOnDevice: ((Int32) -> Bool)? = nil,
        status: ((EpisodeData) -> PlaybackStatus)? = nil
    ) -> Bool {
        if filters.contains(.onDevice), let isOnDevice, !isOnDevice(episode.id) {
            return false
        }
        let dl = downloadChips
        if !dl.isEmpty {
            let ok = dl.contains { chip in
                switch chip {
                case .downloaded: return episode.download_status == .downloaded
                case .downloading: return episode.download_status == .downloading
                default: return false
                }
            }
            if !ok { return false }
        }
        let played = playedChips
        if !played.isEmpty {
            let status = status?(episode) ?? episode.playback_status ?? .unplayed
            let ok = played.contains { chip in
                switch chip {
                case .unplayed: return status == .unplayed
                case .inProgress: return status == .played
                case .finished: return status == .finished
                default: return false
                }
            }
            if !ok { return false }
        }
        return true
    }

    /// Local projection of the same query (the queue's in-memory list).
    func apply(
        to episodes: [EpisodeData], isOnDevice: ((Int32) -> Bool)? = nil,
        status: ((EpisodeData) -> PlaybackStatus)? = nil
    ) -> [EpisodeData] {
        var out = episodes
        let trimmed = search.trimmingCharacters(in: .whitespaces).lowercased()
        if !trimmed.isEmpty {
            out = out.filter {
                $0.title.lowercased().contains(trimmed)
                    || ($0.description?.lowercased().contains(trimmed) ?? false)
                    || ($0.podcast?.title.lowercased().contains(trimmed) ?? false)
            }
        }
        out = out.filter { matchesChips($0, isOnDevice: isOnDevice, status: status) }
        switch orderField {
        case .recency:
            // History sorts recency itself (overlay-wins dates live there);
            // as a plain local sort this is a no-op passthrough.
            break
        case .position:
            // Base order IS position (asc) — desc just flips it (web parity;
            // previously direction was silently ignored for Position).
            if direction == .desc { out.reverse() }
        case .published:
            out.sort {
                direction == .asc
                    ? ($0.published_at ?? .distantPast) < ($1.published_at ?? .distantPast)
                    : ($0.published_at ?? .distantPast) > ($1.published_at ?? .distantPast)
            }
        case .added:
            out.sort {
                direction == .asc
                    ? $0.created_at < $1.created_at : $0.created_at > $1.created_at
            }
        case .title:
            out.sort {
                direction == .asc
                    ? $0.title.lowercased() < $1.title.lowercased()
                    : $0.title.lowercased() > $1.title.lowercased()
            }
        case .duration:
            out.sort {
                direction == .asc
                    ? ($0.duration_secs ?? 0) < ($1.duration_secs ?? 0)
                    : ($0.duration_secs ?? 0) > ($1.duration_secs ?? 0)
            }
        }
        return out
    }
}

/// The persistent sub-navbar on every episode list: search field + Filter
/// dropdown + Order dropdown (the web's list controls, menu-style).
struct ListControlsBar: View {
    @Binding var query: ListQuery
    /// The queue offers Position ordering; other pages don't.
    var allowsPosition = false
    /// History offers the Recent (play-recency) ordering; other pages don't.
    var allowsRecency = false
    /// Downloads fixes its facet via the segmented control — hide the filter
    /// menu there (every other list shows it, History included; web parity).
    var showFilter = true
    /// Embedded accounts have no separate device set — the OnDevice chip is
    /// dropped there (web: `embedded` prop on ListControls).
    var allowsOnDevice = true

    var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 10) {
                HStack(spacing: 6) {
                    Image(systemName: "magnifyingglass")
                        .foregroundStyle(.secondary)
                    TextField("Search", text: $query.search)
                        .textFieldStyle(.plain)
                        .autocorrectionDisabled()
                    if !query.search.isEmpty {
                        Button {
                            query.search = ""
                        } label: {
                            Image(systemName: "xmark.circle.fill")
                                .foregroundStyle(.secondary)
                        }
                        .buttonStyle(.plain)
                    }
                }
                .padding(.horizontal, 10)
                .padding(.vertical, 6)
                .background(RoundedRectangle(cornerRadius: 9).fill(Color(.secondarySystemFill)))

                if showFilter {
                    filterMenu
                }

                Menu {
                    ForEach(orderFields) { field in
                        Button {
                            query.orderField = field
                        } label: {
                            if query.orderField == field {
                                Label(field.label, systemImage: "checkmark")
                            } else {
                                Text(field.label)
                            }
                        }
                    }
                    Divider()
                    Button {
                        query.direction = query.direction == .asc ? .desc : .asc
                    } label: {
                        Label(
                            query.direction == .asc ? "Ascending" : "Descending",
                            systemImage: query.direction == .asc ? "arrow.up" : "arrow.down")
                    }
                } label: {
                    Image(systemName: "arrow.up.arrow.down.circle")
                        .font(.title3)
                }
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 8)
            Divider()
        }
        .background(.bar)
    }

    private var orderFields: [EpisodeOrderField] {
        EpisodeOrderField.allCases.filter {
            (allowsPosition || $0 != .position) && (allowsRecency || $0 != .recency)
        }
    }

    /// Multi-select chips (web ListControls): tapping toggles membership;
    /// the menu stays coherent because facets OR internally / AND across.
    private var filterMenu: some View {
        Menu {
            ForEach(availableChips) { chip in
                Button {
                    if query.filters.contains(chip) {
                        query.filters.remove(chip)
                    } else {
                        query.filters.insert(chip)
                    }
                } label: {
                    if query.filters.contains(chip) {
                        Label(chip.label, systemImage: "checkmark")
                    } else {
                        Text(chip.label)
                    }
                }
            }
            if !query.filters.isEmpty {
                Divider()
                Button("Clear filters") { query.filters = [] }
            }
        } label: {
            Image(
                systemName: query.filters.isEmpty
                    ? "line.3.horizontal.decrease.circle"
                    : "line.3.horizontal.decrease.circle.fill"
            )
            .font(.title3)
        }
    }

    private var availableChips: [EpisodeFilter] {
        EpisodeFilter.allCases.filter { allowsOnDevice || $0 != .onDevice }
    }
}

/// The generic sort+search sub-navbar for NON-episode lists (podcasts,
/// playlists) — the web's shared `SortSearchControls` wrapper: a search field
/// plus a field/direction menu over a page-supplied field vocabulary.
struct SortSearchBar<Field: Hashable>: View {
    @Binding var search: String
    @Binding var field: Field
    @Binding var direction: OrderDirection
    /// Sort fields in menu order: (value, label).
    let fields: [(Field, String)]
    var placeholder = "Search"

    var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 10) {
                HStack(spacing: 6) {
                    Image(systemName: "magnifyingglass")
                        .foregroundStyle(.secondary)
                    TextField(placeholder, text: $search)
                        .textFieldStyle(.plain)
                        .autocorrectionDisabled()
                    if !search.isEmpty {
                        Button {
                            search = ""
                        } label: {
                            Image(systemName: "xmark.circle.fill")
                                .foregroundStyle(.secondary)
                        }
                        .buttonStyle(.plain)
                    }
                }
                .padding(.horizontal, 10)
                .padding(.vertical, 6)
                .background(RoundedRectangle(cornerRadius: 9).fill(Color(.secondarySystemFill)))

                Menu {
                    ForEach(fields.indices, id: \.self) { idx in
                        Button {
                            field = fields[idx].0
                        } label: {
                            if field == fields[idx].0 {
                                Label(fields[idx].1, systemImage: "checkmark")
                            } else {
                                Text(fields[idx].1)
                            }
                        }
                    }
                    Divider()
                    Button {
                        direction = direction == .asc ? .desc : .asc
                    } label: {
                        Label(
                            direction == .asc ? "Ascending" : "Descending",
                            systemImage: direction == .asc ? "arrow.up" : "arrow.down")
                    }
                } label: {
                    Image(systemName: "arrow.up.arrow.down.circle")
                        .font(.title3)
                }
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 8)
            Divider()
        }
        .background(.bar)
    }
}

File: android/app/src/main/kotlin/org/fgsec/halogen/components/ListControls.kt
package org.fgsec.halogen.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.unit.dp
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.SetSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import org.fgsec.halogen.features.latest.EpisodeFilter
import org.fgsec.halogen.networking.WireJson
import org.fgsec.halogen.wire.DownloadStatus
import org.fgsec.halogen.wire.EpisodeData
import org.fgsec.halogen.wire.OrderDirection
import org.fgsec.halogen.wire.PlaybackStatus

/// Sort field vocabulary for episode lists (server `order[order_by]` names).
/// `Position` is the queue's pivot order (local, queue-only).
@Serializable
enum class EpisodeOrderField(val rawValue: String) {
    @SerialName("published_at") Published("published_at"),
    @SerialName("created_at") Added("created_at"),
    @SerialName("title") Title("title"),
    @SerialName("duration_secs") Duration("duration_secs"),
    @SerialName("position") Position("position"),
    /// History only: play-recency order (the web's non-selectable UpdatedAt
    /// sentinel, made selectable) — local, never sent to the server. A
    /// dedicated case so an EXPLICIT "Published desc" choice is
    /// distinguishable from the recency default.
    @SerialName("recency") Recency("recency");

    val label: String
        get() = when (this) {
            Published -> "Published"
            Added -> "Added"
            Title -> "Title"
            Duration -> "Duration"
            Position -> "Position"
            Recency -> "Recent"
        }
}

/// LocalStore keys for the per-page persisted ListQuery — byte-identical to
/// the iOS keys. The owning feature models load/save these, as on iOS.
object ListQueryKeys {
    const val latest = "listquery-latest"
    const val queue = "listquery-queue"
    const val history = "listquery-history"
    const val downloads = "listquery-downloads"
    const val podcastEpisodes = "listquery-podcast-episodes"
    const val playlist = "listquery-playlist"
}

/// One episode list's controls state: search + filter chips + order.
/// Persisted per page (the web's per-view-key sticky list config). Chips are
/// multi-select: OR within a facet, AND across facets (web query.rs).
@Serializable(with = ListQuerySerializer::class)
data class ListQuery(
    val search: String = "",
    val filters: Set<EpisodeFilter> = emptySet(),
    val orderField: EpisodeOrderField = EpisodeOrderField.Published,
    val direction: OrderDirection = OrderDirection.Desc,
) {
    /// Selected chips of each facet.
    val downloadChips: Set<EpisodeFilter>
        get() = filters.intersect(EpisodeFilter.downloadFacet.toSet())
    val playedChips: Set<EpisodeFilter>
        get() = filters.intersect(EpisodeFilter.playedFacet.toSet())

    /// Whether the server page must be re-filtered locally: a facet with
    /// MULTIPLE chips can't be expressed as a single wire token, so the fetch
    /// goes un-narrowed for that facet and `matchesChips` trims each page.
    val needsLocalChipFilter: Boolean
        get() = downloadChips.size > 1 || playedChips.size > 1

    /// Server-side query fragment (search + single-chip facets + order).
    /// Queue/History apply the same query locally instead.
    val queryItems: List<Pair<String, String>>
        get() {
            val items = mutableListOf<Pair<String, String>>()
            val trimmed = search.trim()
            if (trimmed.isNotEmpty()) {
                items += "filter[search]" to trimmed
            }
            downloadChips.singleOrNull()?.wireToken?.let {
                items += "filter[download_status]" to it
            }
            playedChips.singleOrNull()?.wireToken?.let {
                items += "filter[playback_status]" to it
            }
            if (orderField != EpisodeOrderField.Position && orderField != EpisodeOrderField.Recency) {
                items += "order[order_by]" to orderField.rawValue
                items += "order[direction]" to
                    if (direction == OrderDirection.Asc) "Asc" else "Desc"
            }
            return items
        }

    /// Chip predicate over one row: OR within a facet, AND across facets (web
    /// apply_filter_sort). `isOnDevice` null = OnDevice chip no-op (caller switches
    /// sets wholesale); `status` reads the playbacks overlay so local marks flip chips.
    fun matchesChips(
        episode: EpisodeData,
        isOnDevice: ((Int) -> Boolean)? = null,
        status: ((EpisodeData) -> PlaybackStatus)? = null,
    ): Boolean {
        if (filters.contains(EpisodeFilter.OnDevice) && isOnDevice != null &&
            !isOnDevice(episode.id)
        ) {
            return false
        }
        val dl = downloadChips
        if (dl.isNotEmpty()) {
            val ok = dl.any { chip ->
                when (chip) {
                    EpisodeFilter.Downloaded ->
                        episode.download_status == DownloadStatus.Downloaded
                    EpisodeFilter.Downloading ->
                        episode.download_status == DownloadStatus.Downloading
                    else -> false
                }
            }
            if (!ok) return false
        }
        val played = playedChips
        if (played.isNotEmpty()) {
            val resolved = status?.invoke(episode)
                ?: episode.playback_status ?: PlaybackStatus.Unplayed
            val ok = played.any { chip ->
                when (chip) {
                    EpisodeFilter.Unplayed -> resolved == PlaybackStatus.Unplayed
                    EpisodeFilter.InProgress -> resolved == PlaybackStatus.Played
                    EpisodeFilter.Finished -> resolved == PlaybackStatus.Finished
                    else -> false
                }
            }
            if (!ok) return false
        }
        return true
    }

    /// Local projection of the same query (the queue's in-memory list).
    fun apply(
        episodes: List<EpisodeData>,
        isOnDevice: ((Int) -> Boolean)? = null,
        status: ((EpisodeData) -> PlaybackStatus)? = null,
    ): List<EpisodeData> {
        var out = episodes
        val trimmed = search.trim().lowercase()
        if (trimmed.isNotEmpty()) {
            out = out.filter {
                it.title.lowercase().contains(trimmed) ||
                    (it.description?.lowercase()?.contains(trimmed) ?: false) ||
                    (it.podcast?.title?.lowercase()?.contains(trimmed) ?: false)
            }
        }
        out = out.filter { matchesChips(it, isOnDevice, status) }
        return when (orderField) {
            // History sorts recency itself (overlay-wins dates live there);
            // as a plain local sort this is a no-op passthrough.
            EpisodeOrderField.Recency -> out
            // Base order IS position (asc) — desc just flips it (web parity).
            EpisodeOrderField.Position ->
                if (direction == OrderDirection.Desc) out.reversed() else out
            EpisodeOrderField.Published ->
                if (direction == OrderDirection.Asc) out.sortedBy { publishedMillis(it) }
                else out.sortedByDescending { publishedMillis(it) }
            EpisodeOrderField.Added ->
                if (direction == OrderDirection.Asc) out.sortedBy { createdMillis(it) }
                else out.sortedByDescending { createdMillis(it) }
            EpisodeOrderField.Title ->
                if (direction == OrderDirection.Asc) out.sortedBy { it.title.lowercase() }
                else out.sortedByDescending { it.title.lowercase() }
            EpisodeOrderField.Duration ->
                if (direction == OrderDirection.Asc) out.sortedBy { it.duration_secs ?: 0 }
                else out.sortedByDescending { it.duration_secs ?: 0 }
        }
    }

    private fun publishedMillis(episode: EpisodeData): Long =
        episode.published_at
            ?.let { runCatching { WireJson.parseInstant(it).toEpochMilli() }.getOrNull() }
            ?: Long.MIN_VALUE

    private fun createdMillis(episode: EpisodeData): Long =
        runCatching { WireJson.parseInstant(episode.created_at).toEpochMilli() }
            .getOrDefault(Long.MIN_VALUE)
}

/// Lenient per-key decode (the iOS init(from:) shape): pre-multiselect
/// snapshots or a vocabulary change reset the affected field to its default —
/// never fail the whole restore.
object ListQuerySerializer : KSerializer<ListQuery> {
    private val filtersSerializer = SetSerializer(EpisodeFilter.serializer())

    override val descriptor: SerialDescriptor =
        buildClassSerialDescriptor("org.fgsec.halogen.components.ListQuery")

    override fun serialize(encoder: Encoder, value: ListQuery) {
        val out = encoder as JsonEncoder
        out.encodeJsonElement(
            buildJsonObject {
                put("search", out.json.encodeToJsonElement(String.serializer(), value.search))
                put("filters", out.json.encodeToJsonElement(filtersSerializer, value.filters))
                put(
                    "orderField",
                    out.json.encodeToJsonElement(EpisodeOrderField.serializer(), value.orderField))
                put(
                    "direction",
                    out.json.encodeToJsonElement(OrderDirection.serializer(), value.direction))
            })
    }

    override fun deserialize(decoder: Decoder): ListQuery {
        val input = decoder as JsonDecoder
        val obj = input.decodeJsonElement() as? JsonObject ?: return ListQuery()
        fun <T> field(key: String, serializer: KSerializer<T>, default: T): T =
            obj[key]
                ?.let { runCatching { input.json.decodeFromJsonElement(serializer, it) }.getOrNull() }
                ?: default
        return ListQuery(
            search = field("search", String.serializer(), ""),
            filters = field("filters", filtersSerializer, emptySet()),
            orderField = field("orderField", EpisodeOrderField.serializer(), EpisodeOrderField.Published),
            direction = field("direction", OrderDirection.serializer(), OrderDirection.Desc),
        )
    }
}

/// The persistent sub-navbar on every episode list: search field + Filter
/// dropdown + Order dropdown (the web's list controls, menu-style).
@Composable
fun ListControlsBar(
    query: ListQuery,
    onQueryChange: (ListQuery) -> Unit,
    /// The queue offers Position ordering; other pages don't.
    allowsPosition: Boolean = false,
    /// History offers the Recent (play-recency) ordering; other pages don't.
    allowsRecency: Boolean = false,
    /// Downloads fixes its facet via the segmented control — hide the filter
    /// menu there (every other list shows it, History included; web parity).
    showFilter: Boolean = true,
    /// Embedded accounts have no separate device set — the OnDevice chip is
    /// dropped there (web: `embedded` prop on ListControls).
    allowsOnDevice: Boolean = true,
) {
    Column(Modifier.fillMaxWidth().background(MaterialTheme.colorScheme.surfaceContainerLow)) {
        Row(
            Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 8.dp),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            SearchField(
                value = query.search,
                onValueChange = { onQueryChange(query.copy(search = it)) },
                placeholder = "Search",
                modifier = Modifier.weight(1f),
            )
            if (showFilter) {
                FilterMenuButton(
                    query = query, onQueryChange = onQueryChange,
                    allowsOnDevice = allowsOnDevice,
                )
            }
            OrderMenuButton(
                query = query, onQueryChange = onQueryChange,
                allowsPosition = allowsPosition, allowsRecency = allowsRecency,
            )
        }
        HorizontalDivider()
    }
}

/// Multi-select chips (web ListControls): tapping toggles membership; the
/// menu stays coherent because facets OR internally / AND across.
@Composable
private fun FilterMenuButton(
    query: ListQuery,
    onQueryChange: (ListQuery) -> Unit,
    allowsOnDevice: Boolean,
) {
    var open by remember { mutableStateOf(false) }
    val availableChips = EpisodeFilter.entries.filter { allowsOnDevice || it != EpisodeFilter.OnDevice }
    Box {
        IconButton(onClick = { open = true }) {
            Icon(
                halogenIcon(
                    if (query.filters.isEmpty()) "line.3.horizontal.decrease.circle"
                    else "line.3.horizontal.decrease.circle.fill"),
                contentDescription = "Filter",
                tint = if (query.filters.isEmpty()) MaterialTheme.colorScheme.onSurfaceVariant
                else MaterialTheme.colorScheme.primary,
            )
        }
        DropdownMenu(expanded = open, onDismissRequest = { open = false }) {
            for (chip in availableChips) {
                val selected = query.filters.contains(chip)
                DropdownMenuItem(
                    text = { Text(chip.label) },
                    leadingIcon = if (selected) {
                        { Icon(halogenIcon("checkmark"), contentDescription = null) }
                    } else null,
                    onClick = {
                        open = false
                        onQueryChange(
                            query.copy(
                                filters = if (selected) query.filters - chip
                                else query.filters + chip))
                    },
                )
            }
            if (query.filters.isNotEmpty()) {
                HorizontalDivider()
                DropdownMenuItem(
                    text = { Text("Clear filters") },
                    onClick = {
                        open = false
                        onQueryChange(query.copy(filters = emptySet()))
                    },
                )
            }
        }
    }
}

@Composable
private fun OrderMenuButton(
    query: ListQuery,
    onQueryChange: (ListQuery) -> Unit,
    allowsPosition: Boolean,
    allowsRecency: Boolean,
) {
    var open by remember { mutableStateOf(false) }
    val orderFields = EpisodeOrderField.entries.filter {
        (allowsPosition || it != EpisodeOrderField.Position) &&
            (allowsRecency || it != EpisodeOrderField.Recency)
    }
    Box {
        IconButton(onClick = { open = true }) {
            Icon(
                halogenIcon("arrow.up.arrow.down.circle"),
                contentDescription = "Order",
                tint = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        DropdownMenu(expanded = open, onDismissRequest = { open = false }) {
            for (field in orderFields) {
                DropdownMenuItem(
                    text = { Text(field.label) },
                    leadingIcon = if (query.orderField == field) {
                        { Icon(halogenIcon("checkmark"), contentDescription = null) }
                    } else null,
                    onClick = {
                        open = false
                        onQueryChange(query.copy(orderField = field))
                    },
                )
            }
            HorizontalDivider()
            val asc = query.direction == OrderDirection.Asc
            DropdownMenuItem(
                text = { Text(if (asc) "Ascending" else "Descending") },
                leadingIcon = {
                    Icon(
                        halogenIcon(if (asc) "arrow.up" else "arrow.down"),
                        contentDescription = null)
                },
                onClick = {
                    open = false
                    onQueryChange(
                        query.copy(
                            direction = if (asc) OrderDirection.Desc else OrderDirection.Asc))
                },
            )
        }
    }
}

/// The generic sort+search sub-navbar for NON-episode lists (podcasts,
/// playlists) — the web's shared `SortSearchControls` wrapper: a search field
/// plus a field/direction menu over a page-supplied field vocabulary.
@Composable
fun <Field> SortSearchBar(
    search: String,
    onSearchChange: (String) -> Unit,
    field: Field,
    onFieldChange: (Field) -> Unit,
    direction: OrderDirection,
    onDirectionChange: (OrderDirection) -> Unit,
    /// Sort fields in menu order: (value, label).
    fields: List<Pair<Field, String>>,
    placeholder: String = "Search",
) {
    var open by remember { mutableStateOf(false) }
    Column(Modifier.fillMaxWidth().background(MaterialTheme.colorScheme.surfaceContainerLow)) {
        Row(
            Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 8.dp),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            SearchField(
                value = search,
                onValueChange = onSearchChange,
                placeholder = placeholder,
                modifier = Modifier.weight(1f),
            )
            Box {
                IconButton(onClick = { open = true }) {
                    Icon(
                        halogenIcon("arrow.up.arrow.down.circle"),
                        contentDescription = "Order",
                        tint = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
                DropdownMenu(expanded = open, onDismissRequest = { open = false }) {
                    for ((value, label) in fields) {
                        DropdownMenuItem(
                            text = { Text(label) },
                            leadingIcon = if (field == value) {
                                { Icon(halogenIcon("checkmark"), contentDescription = null) }
                            } else null,
                            onClick = {
                                open = false
                                onFieldChange(value)
                            },
                        )
                    }
                    HorizontalDivider()
                    val asc = direction == OrderDirection.Asc
                    DropdownMenuItem(
                        text = { Text(if (asc) "Ascending" else "Descending") },
                        leadingIcon = {
                            Icon(
                                halogenIcon(if (asc) "arrow.up" else "arrow.down"),
                                contentDescription = null)
                        },
                        onClick = {
                            open = false
                            onDirectionChange(if (asc) OrderDirection.Desc else OrderDirection.Asc)
                        },
                    )
                }
            }
        }
        HorizontalDivider()
    }
}

/// The rounded inline search field both bars share (iOS secondarySystemFill
/// capsule): magnifier + plain field + clear button when non-empty.
@Composable
private fun SearchField(
    value: String,
    onValueChange: (String) -> Unit,
    placeholder: String,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier
            .clip(RoundedCornerShape(9.dp))
            .background(MaterialTheme.colorScheme.surfaceContainerHigh)
            .padding(horizontal = 10.dp, vertical = 6.dp),
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(
            halogenIcon("magnifyingglass"),
            contentDescription = null,
            tint = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.size(18.dp),
        )
        Box(Modifier.weight(1f)) {
            if (value.isEmpty()) {
                Text(
                    placeholder,
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            BasicTextField(
                value = value,
                onValueChange = onValueChange,
                singleLine = true,
                textStyle = MaterialTheme.typography.bodyMedium
                    .copy(color = MaterialTheme.colorScheme.onSurface),
                cursorBrush = SolidColor(MaterialTheme.colorScheme.primary),
                modifier = Modifier.fillMaxWidth(),
            )
        }
        if (value.isNotEmpty()) {
            Icon(
                halogenIcon("xmark.circle.fill"),
                contentDescription = "Clear search",
                tint = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.size(18.dp).clickable { onValueChange("") },
            )
        }
    }
}
